File: Mavim-services/Topic/Mavim.Manager.Api.Topic.Services/v1/Mappers/ListFieldMapper.cs
using Mavim.Manager.Api.Topic.Services.v1.Mappers.Abstract;
using Mavim.Manager.Api.Topic.Services.v1.Models.Fields;
using IBusiness = Mavim.Manager.Api.Topic.Business.Interfaces.v1.Fields;
using IService = Mavim.Manager.Api.Topic.Services.Interfaces.v1.Fields;

namespace Mavim.Manager.Api.Topic.Services.v1.Mappers
{
    internal class ListFieldMapper : GenericListFieldMapper<IBusiness.ISingleListField, IService.ISingleListField>
    {
        protected override IService.ISingleListField GetGenericMappedServiceField(IBusiness.ISingleListField field)
        {
            return new SingleListField
            {
                FieldId = field.FieldId,
                FieldName = field.FieldName,
                FieldSetId = field.FieldSetId,
                SetOrder = field.SetOrder,
                Order = field.Order,
                TopicId = field.TopicId,
                SetName = field.SetName,
                Data = field.Data,
                FieldValueType = Map(field.FieldValueType),
                Characteristic = Map(field.Characteristic),
                OpenLocation = field.OpenLocation,
                RelationshipCategory = Map(field.RelationshipCategory),
                Usage = field.Usage,
                Required = field.Required,
                Readonly = field.Readonly,
                Options = field.Options
            };
        }

        protected override IBusiness.ISingleListField GetGenericMappedBusinessField(IService.ISingleListField field)
        {
            return new Business.v1.Models.Fields.SingleListField
            {
                FieldId = field.FieldId,
                FieldSetId = field.FieldSetId,
                SetOrder = field.SetOrder,
                Order = field.Order,
                TopicId = field.TopicId,
                Data = field.Data,
                FieldValueType = Map(field.FieldValueType),
                FieldName = field.FieldName,
                Characteristic = Map(field.Characteristic),
                Required = field.Required,
                Usage = field.Usage,
                Readonly = field.Readonly,
                OpenLocation = field.OpenLocation,
                RelationshipCategory = Map(field.RelationshipCategory),
                SetName = field.SetName,
                Options = field.Options
            };
        }
    }
}

File: Mavim-apps/Mavim.frontend/projects/mavim-app/src/shared/field/mappers/listfield.mapper.ts
import { Field } from '../models/field.model';
import { SingleListField } from '../models/api/';
import { GenericlistFieldMapper } from './abstract';
import { mapToApiFieldType, mapToFieldType } from './fieldvaluetype';

export class ListFieldMapper extends GenericlistFieldMapper<SingleListField> {
	public get getUrlTypeName(): string {
		return 'list';
	}

	public getMappedField(field: SingleListField): Field {
		return this.mapField(field);
	}

	public getMappedApiField(field: Field): SingleListField {
		return this.mapApiField(field);
	}

	protected mapApiField(field: Field): SingleListField {
		return {
			...field,
			options: undefined,
			fieldValueType: mapToApiFieldType(field.fieldValueType),
			data: field.data ? this.mapToDictionary(field.data[0]) : undefined
		} as SingleListField;
	}

	protected mapField(field: SingleListField): Field {
		return {
			...field,
			fieldValueType: mapToFieldType(field.fieldValueType),
			options: this.mapToOptionsString(field.options),
			data: this.mapToString(field.data)
		} as Field;
	}
}
